File: 4031-find-all-numbers-disappeared-in-an-array-ii/4031-find-all-numbers-disappeared-in-an-array-ii.java
class Solution {
    public List<List<Integer>> findDisappearedNumbers(int[] nums, int lower, int upper) {
        
         Set<Integer> number = new HashSet<>();

        for (int num : nums) {
            number.add(num);
        }

        List<Integer> missingInteger = new ArrayList<>();
        for(int i = lower ; i <= upper ; i++){
            if (!number.contains(i)) {
                missingInteger.add(i);
            }
        }

        if(missingInteger.isEmpty()) return new ArrayList<>();

        List<List<Integer>> result = new ArrayList<>();
        int index = 0;

        while(index < missingInteger.size()){
            List<Integer> group = new ArrayList<>();
            group.add(missingInteger.get(index));
            while(index < missingInteger.size() - 1 && missingInteger.get(index) + 1 == missingInteger.get(index + 1)){
                index++;
            }
            group.add(missingInteger.get(index++));
            result.add(new ArrayList<>(group));
        }

        return result;

    }
}

File: 4031-find-all-numbers-disappeared-in-an-array-ii/4031-find-all-numbers-disappeared-in-an-array-ii.py
class Solution:
    def findDisappearedNumbers(self, nums: list[int], lower: int, upper: int) -> list[list[int]]:
        missing = []
        present = set(nums)

        for i in range(lower,upper+1):
            if i not in present:
                missing.append(i)

        if not missing:
            return []
        
        final_missing = []
        start = missing[0]

        for i in range(1 , len(missing)):
            if missing[i] != missing[i-1] + 1:
                final_missing.append([start,missing[i-1]])

                start = missing[i]

        final_missing.append([start,missing[-1]])    
    
        return final_missing
            

        